File: lr/src/questions/q_0290_word_pattern.rs
use std::collections::{hash_map::Entry, HashMap};

pub fn word_pattern(pattern: String, s: String) -> bool {
    let words: Vec<&str> = s.split(' ').collect();
    if pattern.len() != words.len() {
        return false;
    }

    let mut pattern_map = HashMap::new();
    let mut words_map = HashMap::new();

    for (c, word) in pattern.chars().zip(words.iter()) {
        match pattern_map.entry(c) {
            Entry::Occupied(entry) if entry.get() != word => return false,
            Entry::Vacant(entry) => {
                entry.insert(*word);
            }
            _ => {}
        }
        match words_map.entry(*word) {
            Entry::Occupied(entry) if entry.get() != &c => return false,
            Entry::Vacant(entry) => {
                entry.insert(c);
            }
            _ => {}
        }
    }

    true
}

#[test]
fn test_word_pattern() {
    assert!(word_pattern(
        "abba".to_owned(),
        "dog cat cat dog".to_owned()
    ));
    assert!(!word_pattern(
        "abba".to_owned(),
        "dog catt cat dog".to_owned()
    ));
    assert!(!word_pattern(
        "abba".to_owned(),
        "dog cat cat fish".to_owned()
    ));
    assert!(!word_pattern(
        "aaaa".to_owned(),
        "dog cat cat dog".to_owned()
    ));
    assert!(!word_pattern(
        "aaa".to_owned(),
        "cat cat cat dog".to_owned()
    ));
    assert!(!word_pattern("aaa".to_owned(), "cat cat".to_owned()));
    assert!(!word_pattern(
        "abba".to_owned(),
        "dog dog dog dog".to_owned()
    ));
}

File: lt/src/q0290WordPattern.ts
export function wordPattern(pattern: string, s: string): boolean {
    const words = s.split(" ");
    if (pattern.length !== words.length) {
        return false;
    }
    const patternMap: Map<string, number> = new Map();
    const wordsMap: Map<string, number> = new Map();

    for (let index = 0; index < pattern.length; index++) {
        const word = words[index];
        const c = pattern[index];
        if (!patternMap.has(c)) {
            patternMap.set(c, index);
        }
        if (!wordsMap.has(word)) {
            wordsMap.set(word, index);
        }

        if (patternMap.get(c) !== wordsMap.get(word)) {
            return false;
        }
    }

    return true;
}
